staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -260,22 +286,45 @@ func (t *envelopeTransformer) doEncode(request *kmstypes.EncryptedObject) ([]byt
 	if err := ValidateEncryptedObject(request); err != nil {
 		return nil, err
 	}
-	return proto.Marshal(request)
+
+	encodingBuffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
+	defer t.encryptedObjectDecoder.bufferPool.Put(encodingBuffer)
+
+	encodingBuffer.Reset()
+	if err := encodingBuffer.Marshal(request); err != nil {
+		return nil, err
+	}
+
+	return encodingBuffer.Bytes(), nil
```

Comment:
This `defer` and `encodingBuffer.Bytes()` are not compatible with each other - `Bytes()` just returns the underlying slice which could be re-used before the caller is done with it.  We have a test gap here because nothing is failing.

```suggestion
	encodingBuffer := t.encryptedObjectDecoder.bufferPool.Get().(*proto.Buffer)
	defer t.encryptedObjectDecoder.bufferPool.Put(encodingBuffer)

	encodingBuffer.Reset()
	if err := encodingBuffer.Marshal(request); err != nil {
		return nil, err
	}

	return bytes.Clone(encodingBuffer.Bytes()), nil
```